### PR TITLE
add Half support for flash attention on CPU

### PR DESCRIPTION
**Pull request by CaoE**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #118368
* #118367

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

---
Appease ShipIt:
@diff-train-skip-merge

### CODE REVIEW

**Review by jgong5 — APPROVED** [review]

**Review by Valentine233 — APPROVED** [review]

**Comment by Valentine233** [comment]
Could you add perf data comparisons for models like Stable Diffusion and Bert?

**Comment by CaoE** [comment]
@Valentine233 The performance data of FP16 requires the 5th generation Intel® Xeon® Scalable processor. This PR is to add Half support for flash attention on CPU, which can also improve accuracy and avoid overflows for FP16 data type since `_scaled_dot_product_attention_math` use FP16 as intermediate results.

**Comment by CaoE** [comment]
@drisspg Could you please help review this PR ?

**Review by drisspg — APPROVED** [review]
Looks good, thank you!

**Comment by clee2000** [comment]
Hi sorry I reverted this in https://github.com/pytorch/pytorch/pull/119204 because it's breaking some internal builds.  I put a copy of the stack trace in the revert PR

**Comment by drisspg** [comment]
@CaoE  I think you need to update the decomp table:
this line 
https://github.com/pytorch/pytorch/blob/main/torch/_decomp/decompositions.py#L4322-L4325


needs to be updated to: 

```Python
	
    torch._check(
        torch.is_floating_point(query),
        lambda: f"query must be FP32, FP64, BF16, FP16 but got {query.dtype}",
    )
 ```

**Comment by clee2000** [comment]
Could you remake these changes in a new PR?  The auto branch deletion is preventing us from reopening so it can't be changed or remerged

**Comment by CaoE** [comment]
@drisspg Thanks for your comments. Fixed in https://github.com/pytorch/pytorch/pull/119247.

**Comment by mikekgfb** [comment]
> This pull request is closed, but the gh/CaoE/42/head branch has unmerged commits.

Hi @CaoE , I did all your changes land successfully?  Or did I land an earlier version and missed some of your changes?  (We needed this for some internal testing so we merged/landed it very quickly.)

**Comment by CaoE** [comment]
@mikekgfb Sorry for the late reply, as I am currently on holiday.

> I did all your changes land successfully? 

https://github.com/pytorch/pytorch/pull/119247 landed all changes.

**Comment by CaoE** [comment]
@drisspg We found that nightly torch is built with MKL 2022.0. MKL >= 2024.0 support the latest fp16 feature. Currently, fp16 sdpa will be very slow because it can only use the fallback gemm. Will torch consider using the latest MKL? May I know if there are any schedule ?

`[Info] Running torch.compile() Float32 with default backend
MKL_VERBOSE oneMKL 2022.0 Update 2 Product build 20220804 for Intel(R) 64 architecture Intel(R) Advanced Vector Extensions 512 (Intel(R) AVX-512) with Intel(R) DL Boost, bfloat16 support, Intel(R) AMX with bfloat16 and 8-bit integer support, and FP16 instructions, Lnx 0.90GHz lp64 gnu_thread`

**Comment by drisspg** [comment]
@CaoE can you create a separate issue for upgrading MKL?

**Comment by CaoE** [comment]
@drisspg Thanks for your comments. Create a new issue for this https://github.com/pytorch/pytorch/issues/121201.
